Assignment3/app.py

```diff
@@ -200,12 +198,14 @@ def _ReceiveMessages(self):
 
     # Send data to the other party
     def _SendMessage(self, message):
-        print("PRE ENCRYPT: sending the message:")
-        print(message)
+        if not self.prtcl._session_key and not self.prtcl.IsMessagePartOfProtocol(message):
```

Comment:
_session_key is a private variable and we should not be accessing it outside of protocol class.

You could check for is_ping_pong_done and its a boolean.